sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraIO.java

```diff
@@ -233,4 +257,136 @@ public void populateDisplayData(DisplayData.Builder builder) {
+
+    public Write<T> withCassandraService(CassandraService<T> cassandraService) {
+      checkArgument(cassandraService != null, "CassandraIO.write().withCassandraService"
+          + "(service) called with null service");
+      return builder().setCassandraService(cassandraService).build();
+    }
+
+    @Override
+    public void validate(PCollection<T> input) {
+      checkState(hosts() != null || cassandraService() != null,
+          "CassandraIO.write() requires a list of hosts to be set via withHosts(hosts) or a "
+              + "Cassandra service to be set via withCassandraService(service)");
+      checkState(port() != null || cassandraService() != null, "CassandraIO.write() requires a "
+          + "valid port number to be set via withPort(port) or a Cassandra service to be set via "
+          + "withCassandraService(service)");
+      checkState(keyspace() != null, "CassandraIO.write() requires a keyspace to be set via "
+          + "withKeyspace(keyspace)");
+      checkState(entity() != null, "CassandraIO.write() requires an entity to be set via "
+          + "withEntity(entity)");
+    }
+
+    @Override
+    public PDone expand(PCollection<T> input) {
+      input.apply(ParDo.of(new WriteFn<T>(this,
+          new SerializableFunction<PipelineOptions, CassandraService<T>>() {
+        @Override
+        public CassandraService<T> apply(PipelineOptions input) {
+          return getCassandraService(input);
+        }
+      })));
+      return PDone.in(input.getPipeline());
+    }
+
+    @AutoValue.Builder
+    abstract static class Builder<T> {
+      abstract Builder<T> setHosts(List<String> hosts);
+      abstract Builder<T> setPort(Integer port);
+      abstract Builder<T> setKeyspace(String keyspace);
+      abstract Builder<T> setEntity(Class<T> entity);
+      abstract Builder<T> setCassandraService(CassandraService<T> cassandraService);
+      abstract Write<T> build();
+    }
+
+    /**
+     * Helper function to either get a fake/mock Cassandra service provided by
+     * {@link #withCassandraService(CassandraService)} or creates and returns an implementation
+     * of a concrete Cassandra service dealing with a Cassandra instance.
+     */
+    @VisibleForTesting
+    CassandraService<T> getCassandraService(PipelineOptions pipelineOptions) {
+      if (cassandraService() != null) {
+        return cassandraService();
+      }
+      return new CassandraServiceImpl<>();
+    }
+
+  }
+
+  private static class WriteFn<T> extends DoFn<T, Void> {
+
+    private final Write<T> spec;
+    private final SerializableFunction<PipelineOptions, CassandraService<T>>
+        cassandraServiceFactory;
+    private CassandraService.Writer writer;
+
+    public WriteFn(Write<T> spec,
+                   SerializableFunction<PipelineOptions, CassandraService<T>>
+                       cassandraServiceFactory) {
+      this.spec = spec;
+      this.cassandraServiceFactory = cassandraServiceFactory;
+    }
+
+    @StartBundle
```

Comment:
You may want to do this in `@Setup` instead. Bundles can be very small in streaming.